### PR TITLE
Array interface

### LINKED ISSUE #2935: __array__ and __array_wrap__
To improve Numpy interoperability, torch tensors could implement the array interface in a quite simple manner:
```python
def __array__(self, dtype=None):
    if dtype is None:
        return self.cpu().numpy()
    else:
        return self.cpu().numpy().astype(dtype, copy=False)


def __array_wrap__(self, array):
    if array.ndim == 0:
        if array.dtype.kind == 'b':
            return bool(array)
        elif array.dtype.kind in ('i', 'u'):
            return int(array)
        elif array.dtype.kind == 'f':
            return float(array)
        elif array.dtype.kind == 'c':
            return complex(array)
        else:
            raise RuntimeError
    else:
        if array.dtype == bool:
            # Workaround, torch has no built-in bool tensor
            cls = torch.ByteTensor
            array = array.astype('uint8')
        else:
            cls = _tensor_cls(array.dtype, use_cuda=self.is_cuda)
        return cls(array)


def _tensor_cls(dtype, use_cuda):
    mapping = {
        np.dtype('float16'): 'Half',
        np.dtype('float32'): 'Float',
        np.dtype('float64'): 'Double',
        np.dtype('int8'): 'Char',
        np.dtype('int16'): 'Short',
        np.dtype('int32'): 'Int',
        np.dtype('int64'): 'Long',
        np.dtype('uint8'): 'Byte',
    }
    try:
        name = mapping[np.dtype(dtype)]
    except KeyError:
        raise ValueError('dtype {} not supported'.format(dtype))
    if use_cuda:
        return getattr(torch.cuda, name)
    else:
        return getattr(torch, name)
```
It would enable simple things like `numpy.asarray` to turn tensors into Numpy arrays, and with that make basically all of Numpy applicable to torch tensors (with some restrictions). Is something like that of interest?

### PR DESCRIPTION
**Pull request by kohr-h**
Closes #2935

### CODE REVIEW

File: torch/__init__.py [hunk 1]
```diff
@@ -185,6 +185,36 @@ class ByteStorage(_C.ByteStorageBase, _StorageBase):
...
+        elif array.dtype.kind == 'c':
+            return complex(array)
+        else:
+            raise RuntimeError('bad scalar {!r}'.format(array))
+    else:
+        if array.dtype == bool:
+            # Workaround, torch has no built-in bool tensor
+            array = array.astype('uint8')
+
+        return from_numpy(array)
```
Comment (apaszke) [inline torch/__init__.py]: You can add these two to `_TensorBase`

**Review by apaszke — COMMENTED** [review]
This looks great! I left a few comments

**Comment by fmassa** [comment]
Do we also want to add this to `Variable`?

**Comment by fmassa** [comment]
The downside of adding it to `Variable` is that it could lead to confusion to the user, thinking that because doing `np.sin(variable)` works, then autograd should work on it.
But with the unification of `Variable` and `Tensor`, I think we should add it anyway.

File: torch/__init__.py [hunk 2]
```diff
@@ -185,6 +185,36 @@ class ByteStorage(_C.ByteStorageBase, _StorageBase):
...
+    if dtype is None:
+        return self.cpu().numpy()
+    else:
+        return self.cpu().numpy().astype(dtype, copy=False)
+
+
+# Wrap Numpy array again in a suitable tensor when done, to support e.g.
+# `numpy.sin(tensor) -> tensor` or `numpy.greater(tensor, 0) -> ByteTensor`
+def __array_wrap__(self, array):
+    if array.ndim == 0:
```
Comment (fmassa) [inline torch/__init__.py]: for later: we will be adding support for 0-dim tensors in PyTorch at some point, so maybe it would be good to add a TODO here to modify this once it's implemented?

File: torch/__init__.py (hunk 1 above)
Comment (kohr-h) [reply torch/__init__.py]: Sounds easier, I'll do that.

File: test/test_torch.py [hunk 3]
```diff
@@ -4231,6 +4231,83 @@ def test_numpy_index(self):
...
 
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_numpy_array_interface(self):
+        types = [
+            'torch.DoubleTensor',
+            'torch.FloatTensor',
+            'torch.LongTensor',
+            'torch.IntTensor',
+            'torch.ShortTensor',
+            'torch.ByteTensor',
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply test/test_torch.py]: Will do

**Comment by kohr-h** [comment]
> Do we also want to add this to Variable?

Since I don't have an opinion on that, I'll wait for more comments.

File: torch/__init__.py (hunk 2 above)
Comment (kohr-h) [reply torch/__init__.py]: Sure, I'll add a TODO.

**Comment by kohr-h** [comment]
Comments addressed.

**Comment by apaszke** [comment]
Let's not add them to Variables yet. We'll do this later, after we remove stochastic functions and stop creating graphs when requires_grad is False.

File: torch/__init__.py (hunk 1 above)
Comment (apaszke) [reply torch/__init__.py]: That's not what I meant. Add them as regular methods [here](https://github.com/pytorch/pytorch/blob/master/torch/tensor.py#L8).

Comment (kohr-h) [reply torch/__init__.py]: Ah, alright. Didn't realize that `_TensorBase` was implemented in Python. I'll move the code.

**Comment by kohr-h** [comment]
Ok, code moved.

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Build failures seem to be caused by an error on our side. Still, I'll wait until they are fixed to confirm that this patch is correct.

**Comment by soumith** [comment]
thank you!

**Comment by kohr-h** [comment]
My pleasure
